src/gui/input.cpp

```diff
@@ -1,263 +1,224 @@
-		}
-    }
+	static const QMap<int, QString> keypadKeys {
+		{ Qt::Key_Home, "<%1kHome>" },
+		{ Qt::Key_End, "<%1kEnd>" },
+		{ Qt::Key_PageUp, "<%1kPageUp>" },
+		{ Qt::Key_PageDown, "<%1kPageDown>" },
+		{ Qt::Key_Plus, "<%1kPlus>" },
+		{ Qt::Key_Minus, "<%1kMinus>" },
+		{ Qt::Key_multiply, "<%1kMultiply>" },
+		{ Qt::Key_division, "<%1kDivide>" },
+		{ Qt::Key_Enter, "<%1kEnter>" },
+		{ Qt::Key_Period, "<%1kPoint>" },
+		{ Qt::Key_0, "<%1k0>" },
+		{ Qt::Key_1, "<%1k1>" },
+		{ Qt::Key_2, "<%1k2>" },
+		{ Qt::Key_3, "<%1k3>" },
+		{ Qt::Key_4, "<%1k4>" },
+		{ Qt::Key_5, "<%1k5>" },
+		{ Qt::Key_6, "<%1k6>" },
+		{ Qt::Key_7, "<%1k7>" },
+		{ Qt::Key_8, "<%1k8>" },
+		{ Qt::Key_9, "<%1k9>" },
+	};
+
+	if (mod & Qt::KeypadModifier && keypadKeys.contains(key)) {
+		return keypadKeys.value(key).arg(GetModifierPrefix(mod));
+	}
+
+	const QMap<int, QString>& specialKeys { GetSpecialKeysMap() };
 
-	if (specialKeys.contains(k)) {
-		return QString("<%1%2>").arg(modPrefix(mod)).arg(specialKeys.value(k));
+	if (specialKeys.contains(key)) {
+		return ToKeyString(GetModifierPrefix(mod), specialKeys.value(key));
 	}
 
-	QChar c;
-	// Escape < and backslash
+	// The key "<" should be sent as "<lt>"
+	// Issue#607: Remove ShiftModifier from "<", shift is implied.
 	if (text == "<") {
-		return QString("<%1%2>").arg(modPrefix(mod)).arg("lt");
-	} else if (text == "\\") {
-		return QString("<%1%2>").arg(modPrefix(mod)).arg("Bslash");
-	} else if (text.isEmpty()) {
-		// on macs, text is empty for ctrl+key and cmd+key combos (with or without alt)
-		if (mod & ControlModifier || mod & CmdModifier) {
-			// ignore ctrl, alt and cmd key combos by themselves
-			QList<Qt::Key> keys = { Key_Control, Key_Alt, Key_Cmd };
-			if (keys.contains((Qt::Key)k)) {
-				return QString();
-			} else if (mod & ShiftModifier) {
-				// Ignore event for Ctrl-Shift
-				// Fixes issue #344, C-S- being treated as C-Space
-				return QString();
-			} else {
-				// key code will be the value of the char (hopefully)
-				c = QChar(k);
-			}
-		} else {
+		const auto modNoShift { mod & ~Qt::KeyboardModifier::ShiftModifier };
+		return ToKeyString(GetModifierPrefix(modNoShift), "lt");
+	}
+
+	if (text == "\\") {
+		return ToKeyString(GetModifierPrefix(mod), "Bslash");
+	}
+
+	if (text.isEmpty()) {
+		if (!(mod & ControlModifier()) && !(mod & CmdModifier())) {
 			// This is a special key we can't handle
-			return QString();
+			return {};
+		}
+
+		// Ignore all modifier-only key events.
+		if (key == Qt::Key::Key_Alt
+			|| key == Qt::Key::Key_Control
+			|| key == Qt::Key::Key_Meta
+			|| key == Qt::Key::Key_Super_L
+			|| key == Qt::Key::Key_Super_R) {
+			return {};
+		}
+
+		// Issue #344: Ignore Ctrl-Shift, C-S- being treated as C-Space
+		if (mod & Qt::KeyboardModifier::AltModifier
+			|| mod & Qt::KeyboardModifier::ControlModifier
+			|| mod & Qt::KeyboardModifier::MetaModifier
+			|| mod & Qt::KeyboardModifier::ShiftModifier) {
+			return {};
```

Comment:
This is breaking things like `^w` on macOS Catalina. For some reason, when `^w` is pressed (or most any key with control or command), `text` is empty, but `key` is still `87` (the valid ASCII code). Adding a check for `if (key > 255 && ...` fixes the issue without re-introducing #344, but I've no clue how portable that solution is.

Reply:
I don't think the `key > 255` is safe for international keyboard layouts. I think it is possible to have Ctrl + {Non English Character} combinations, but I am not 100% sure.

Can you re-test if I revert this snippet?

Reply:
Ready for re-test :)

If you can get me the relevant QKeyEvents too, that would be most appreciated!

Reply:
So this gets basic combos back, but now any keypress involving shift gets ignored:

```
// w key down
QKeyEvent ev: QKeyEvent(KeyPress, Key_W, text="w")
sent to vim: 'w'
// ctrl key down
QKeyEvent ev: QKeyEvent(KeyPress, Key_Meta, MetaModifier)
sent to vim: ''
// ctrl+w
QKeyEvent ev: QKeyEvent(KeyPress, Key_W, MetaModifier)
sent to vim: '<C-W>'
// ctrl key down
QKeyEvent ev: QKeyEvent(KeyPress, Key_Meta, MetaModifier)
sent to vim: ''
// shift key down
QKeyEvent ev: QKeyEvent(KeyPress, Key_Shift, ShiftModifier|MetaModifier)
sent to vim: ''
// ctrl+shift+w
QKeyEvent ev: QKeyEvent(KeyPress, Key_W, ShiftModifier|MetaModifier)
sent to vim: ''
```

I wonder if even ever using `ev->text()` is the right thing to do. On macOS, `alt+x` gives the unicode symbol `≈`, so they get sent to vim as `<A-≈>`. Whether it should be sent as simply `≈` or `<A-x>` is debatable.

Reply:
And if it helps, here's what `ctrl+shift` looks like (I removed the `if` block checking just to show what would naively be sent to vim)

```
// ctrl key down
QKeyEvent ev: QKeyEvent(KeyPress, Key_Meta, MetaModifier)
sent to vim: ''
// ctrl+shift
QKeyEvent ev: QKeyEvent(KeyPress, Key_Shift, ShiftModifier|MetaModifier)
sent to vim: '<C- >'
// shift key down
QKeyEvent ev: QKeyEvent(KeyPress, Key_Shift, ShiftModifier)
sent to vim: ''
// shift+ctrl
QKeyEvent ev: QKeyEvent(KeyPress, Key_Meta, ShiftModifier|MetaModifier)
sent to vim: ''
```

looks like there's a difference somehow, depending on which key is pressed first.

Reply:
Also seems like this is a known thing in Qt:

https://doc.qt.io/qt-5/qkeyevent.html#text

> Return values when modifier keys such as Shift, Control, Alt, and Meta are pressed differ among platforms and could return an empty string.

Reply:
> I wonder if even ever using ev->text() is the right thing to do. On macOS, alt+x gives the unicode symbol ≈, so they get sent to vim as <A-≈>. Whether it should be sent as simply ≈ or <A-x> is debatable.

The scenario you mention is essentially the same as #579 :)

I would lean towards letting the platform escape the character... You can disable this behavior by setting you keyboard to Unicode Hex in MacOS. I think there is a good argument for either behavior. How can we identify user intent?

Reply:
> So this gets basic combos back, but now any keypress involving shift gets ignored:
...
// ctrl+shift+w
QKeyEvent ev: QKeyEvent(KeyPress, Key_W, ShiftModifier|MetaModifier)
sent to vim: ''

Does this work in master? Based on a read of the code in master, I suspect no. The code for `text.isEmpty()` should be equivalent now... Could be a mistake on my part to.

If this works in master, we have a regression. If this doesn't, we have a long-standing bug. We can fix it either way.

Sorry, will need some help. I can't test natively on MacOS for a while...

Reply:
IMO... This  is all too difficult to reason through manually. There are just too many corner cases based on platform, localization, and keyboard layout. That is why there are so many input bugs right now.

If we can write reliable unit tests for all of these scenarios, we ensure that each case is handled properly and we don't regress things.

Reply:
> I think there is a good argument for either behavior. How can we identify user intent?

I'm really not sure here. I can see arguments for both ways: vim heavy users might expect `<M-x>` to come through, while "a good macOS citizen" might expect `≈`. Either way, what gets sent from nvim-qt (`<M-≈>`), seems to be the oddest.

For the record:

- macvim sends `≈`
- master sends `<M-≈>`
- I'd prefer it send `<M-x>`. While it's less "macOS", I think it's more intuitive, as `<M-x>` is `<M-≈>` while `<M-S-X>` is `<M-˛>`

> If this works in master, we have a regression. If this doesn't, we have a long-standing bug.

Control-shift / command-shift mappings don't work in master either, it's a long-standing issue. This PR made it much easier for me to track down, and that's appreciated.

> Sorry, will need some help.

No worries, I am happy to help out where I can. It's been a little hectic here due to holidays, but that's mostly over for now.

Reply:
What about:

```cpp
		if (
			(
				key == Qt::Key::Key_Shift
				|| key == Key_Control()
			)
			&& (mod & Qt::KeyboardModifier::ShiftModifier)
			&& (mod & ControlModifier())
		) {
                }
```

That seems to stop #344 from happening on macOS, while not blocking anything else from going through.

I can't test on Windows, which sucks cause that was the platform of the original issue, but I can test on Linux.

As for the other issue I mentioned, I don't think it should hold this work back.

Reply:
Thanks for the help, and happy holidays! Likewise, I may be in and out over the next few days...

I can handle Linux/Windows testing. But, MacOS is much less familiar and available to me.

With the current changes 344 is working again, no?

The simplest fix for 344 is probably:
``` patch
                if (key == Qt::Key::Key_Alt
                        || key == Qt::Key::Key_Control
                        || key == Qt::Key::Key_Meta
+                      || key == Qt::Key::Key_Shift
```

This should be functionally equivalent because:
`Key_Shift` implies mod & Qt::ShiftModifier
`Key_Control` implies mod & ControlModifier()

Regarding 579, more work would be required. I'm also not sure how this would interact with Issue 510 (pinging affected users).

I suspect to properly fix the issue, we would need to ignore `QKeyEvent::text()` and do something similar to `GetSpecialKeysMap()` to map a `Qt::Key` to a corresponding string.

Reply:
> With the current changes 344 is working again, no?

Correct. Sorry if I what I said was confusing. I should have clarified that #344 is *still* fixed, in addition to other mappings now working on mac.

> The simplest fix for 344 is probably

Can confirm this works. Now, given that it's only checking `key`, can this then be moved near the top of `Input::convertKey`?

> I suspect to properly fix the issue, we would need to ignore QKeyEvent::text() and do something similar to GetSpecialKeysMap() to map a Qt::Key to a corresponding string.

That's what I'm thinking. I've tried a couple of things, both work:

- platform specific `QKeyEvent Input::NormalizeQKeyEvent(QKeyEvent *ev)`
- platform specific `QString Input::GetTextFromQKeyEvent(QKeyEvent *ev)`

They both boil down to the same thing on macos:

```cpp
QString GetTextFromQKeyEvent(QKeyEvent *ev) noexcept
{
	const QString text = ev->text();
	const int key = ev->key();
	const Qt::KeyboardModifiers modifiers = ev->modifiers();

        // Convert macOS Alt+ mappings back to their normal characters.
	if (
		!text.isEmpty()
		&& modifiers & Qt::KeyboardModifier::AltModifier
		&& !(
			modifiers & CmdModifier()
			&& modifiers & ControlModifier()
		)
	) {
		return QString(key);
	}

        // On macOS, QKeyEvent::text() is empty when pressing control or command.
        // https://doc.qt.io/qt-5/qkeyevent.html#text
	if (
		text.isEmpty()
		&& (
			modifiers & ControlModifier()
			|| modifiers & CmdModifier()
		)
	) {
		return QString(key);
	}

	return text;
}
```

Calling that where appropriate in `shell.cpp` fixes two oddities on macOS:

- Things like `Alt+X` get mapped to `M+X` now, instead of `M-≈`
- `text` is no longer empty when command or control are pressed on macos. That seems unexpected, and what was causing these issues with mapping `shift` on mac.

But honestly, this works as well:

```cpp
QString GetTextFromQKeyEvent(QKeyEvent *ev) noexcept
{
	const QChar c = QChar(ev->key());

	if (c.isPrint()) {
		if (!(ev->modifiers() & Qt::KeyboardModifier::ShiftModifier)) {
			return c.toLower();
		}

		return c;
	}

	return "";
}
```

Which just makes me think `QKeyEvent::text()` is dubious.

---

So there's a few ways of solving the `Alt+` mappings on macOS. Both the long and short versions work the same on all internation keyboards I tested (U.S, French, German - Standard, Belarusian).

I've no idea about AltGr, so can't weigh in on #510. But I do know things get extra hairy: nvim-qt will catch `Alt+E` and people can map it, but that puts macOS into accent aigu mode. So if the *does* map it, then they next type `a`, they'll get `á`.

Reply:
Since 344 works again, I am going to resolve this thread.

Let's move the MacOS input normalization discussion over to #579. I think the fix will just complicate and slow down this Pull Request. It will be easier to review and validate small incremental changes.

I am still a little hesitant on this, `M-≈` is technically valid input according to the neovim input specifications.

> Can confirm this works.

Let's go with the simpler fix. I will force-push the `Qt::Key::Key_Shift` changes.

> Now, given that it's only checking key, can this then be moved near the top of Input::convertKey?

I would prefer to leave as-is. I think it is a good idea to keep the code similar to what was there before. This makes review and the diff/history easier to read. If I were writing from scratch, I would definitely do this.

>That's what I'm thinking. I've tried a couple of things, both work:
>
> - platform specific QKeyEvent Input::NormalizeQKeyEvent(QKeyEvent *ev)
> - platform specific QString Input::GetTextFromQKeyEvent(QKeyEvent *ev)
I prefer a `GetTextFromQKeyEvent` type solution. I don't think we should modify the QKeyEvent sent by the platform.

Of the two functions you provided, I prefer the first example.

I think `toLower()` is a code smell, and I prefer to avoid it. While safe for English, not all language characters map cleanly between cases. This can lead to some unexpected bugs...